integration-tests/src/test_loop/utils/sharding.rs

```diff
@@ -0,0 +1,155 @@
+use borsh::BorshDeserialize;
+use near_chain::types::Tip;
+use near_chain::ChainStoreAccess;
+use near_client::Client;
+use near_epoch_manager::EpochManagerAdapter;
+use near_primitives::hash::CryptoHash;
+use near_primitives::shard_layout::ShardLayout;
+use near_primitives::state_record::StateRecord;
+use near_primitives::types::ShardId;
+use near_store::adapter::StoreAdapter;
+use near_store::db::refcount::decode_value_with_rc;
+use near_store::{DBCol, ShardUId, Trie};
+
+/// Returns `true` if `client` is tracking the shard having the given `shard_id`.
+pub fn client_tracking_shard(client: &Client, shard_id: ShardId, parent_hash: &CryptoHash) -> bool {
+    let signer = client.validator_signer.get();
+    let account_id = signer.as_ref().map(|s| s.validator_id());
+    client.shard_tracker.care_about_shard(account_id, parent_hash, shard_id, true)
+}
+
+// Finds the client who tracks the shard with `shard_id` among the list of `clients`.
+pub fn get_client_tracking_shard<'a>(
+    clients: &'a [&Client],
+    tip: &Tip,
+    shard_id: ShardId,
+) -> &'a Client {
+    for client in clients {
+        if client_tracking_shard(client, shard_id, &tip.prev_block_hash) {
+            return client;
+        }
+    }
+    panic!(
+        "get_client_tracking_shard() could not find client tracking shard {} at {} #{}",
+        shard_id, &tip.last_block_hash, tip.height
+    );
+}
+
+/// Prints the accounts inside all shards and asserts that no shard is empty.
+pub fn print_and_assert_shard_accounts(clients: &[&Client], tip: &Tip) {
+    let epoch_config = clients[0].epoch_manager.get_epoch_config(&tip.epoch_id).unwrap();
+    for shard_uid in epoch_config.shard_layout.shard_uids() {
+        let client = get_client_tracking_shard(clients, tip, shard_uid.shard_id());
+        let chunk_extra = client.chain.get_chunk_extra(&tip.prev_block_hash, &shard_uid).unwrap();
+        let trie = client
+            .runtime_adapter
+            .get_trie_for_shard(
+                shard_uid.shard_id(),
+                &tip.prev_block_hash,
+                *chunk_extra.state_root(),
+                false,
+            )
+            .unwrap();
+        let mut shard_accounts = vec![];
+        for item in trie.lock_for_iter().iter().unwrap() {
+            let (key, value) = item.unwrap();
+            let state_record = StateRecord::from_raw_key_value(key, value);
+            if let Some(StateRecord::Account { account_id, .. }) = state_record {
+                shard_accounts.push(account_id.to_string());
+            }
+        }
+        println!("accounts for shard {}: {:?}", shard_uid, shard_accounts);
+        assert!(!shard_accounts.is_empty());
+    }
+}
+
+/// Asserts that all parent shard State is accessible via parent and children shards.
+pub fn check_state_shard_uid_mapping_after_resharding(client: &Client, parent_shard_uid: ShardUId) {
```

Comment:
I would move this function to `trie_sanity`.

Reply:
Done